dom/traverse.ts

```diff
@@ -59,91 +59,91 @@ function fetchSingleSibling (element : HTMLElement, selector : null|string, acce
 /**
  * Finds all DOM elements matching the selector
  */
-export function find (selector : string, context : HTMLElement|Document = document) : HTMLElement[]
+export function find<T extends HTMLElement = HTMLElement> (selector : string, context : HTMLElement|Document = document) : T[]
 {
-    return Array.prototype.slice.call(context.querySelectorAll(selector));
+    return Array.prototype.slice.call(context.querySelectorAll(selector)) as T[];
 }
 
 
 /**
  * Finds a single DOM node matching the selector
  */
-export function findOne (selector : string, context : HTMLElement|Document = document) : null|HTMLElement
+export function findOne<T extends HTMLElement = HTMLElement> (selector : string, context : HTMLElement|Document = document) : null|T
 {
-    return context.querySelector(selector);
+    return context.querySelector(selector) as T;
 }
 
 
 /**
  * Filters a list of DOM elements that match the given selector
  */
-export function filter (list : HTMLElement[], selector : string) : HTMLElement[]
+export function filter<T extends HTMLElement = HTMLElement> (list : HTMLElement[], selector : string) : T[]
 {
     return list.filter(
         (e) => e.matches(selector)
-    );
+    ) as T[];
 }
 
 
 /**
  * Filters a list of DOM elements that DO NOT match the given selector,
  * are not the given node or are not in the given node list.
  */
-export function not (list : HTMLElement[], selector : string|HTMLElement|HTMLElement[]) : HTMLElement[]
+export function not<T extends HTMLElement = HTMLElement> (list : HTMLElement[], selector : string|HTMLElement|HTMLElement[]) : T[]
```

Comment:
I would write it differently. Basically the return type is a subset of `list`, so the specific types of `list` should just be passed through without specifying them manually.

Reply:
selector needs no specialized types, as it is only used internally and the specific type doesn't matter.

If you pass in a mixed set of elements, then your type is `HTMLElement[]` before and will stay it as it is. But the function preserves the type of the parameter and we should describe it as such.

Reply:
You can just do:

```typescript
xport function not<TList extends HTMLElement[]> (list: TList, selector: string|HTMLElement|HTMLElement[]): TList
{
    // ...
}